src/biz/workspace/publish.rs

```diff
@@ -154,3 +120,296 @@ async fn check_workspace_namespace(new_namespace: &str) -> Result<(), AppError>
+    delete_published_collabs(&self.pg_pool, workspace_id, view_ids).await?;
+    Ok(())
+  }
+}
+
+pub struct PublishedCollabS3StoreWithPostgresFallback {
+  metrics: Arc<PublishedCollabMetrics>,
+  pg_pool: PgPool,
+  bucket_client: AwsS3BucketClientImpl,
+}
+
+impl PublishedCollabS3StoreWithPostgresFallback {
+  pub fn new(
+    metrics: Arc<PublishedCollabMetrics>,
+    pg_pool: PgPool,
+    bucket_client: AwsS3BucketClientImpl,
+  ) -> Self {
+    Self {
+      metrics,
+      pg_pool,
+      bucket_client,
+    }
+  }
+}
+
+#[async_trait]
+impl PublishedCollabStore for PublishedCollabS3StoreWithPostgresFallback {
+  async fn publish_collabs(
+    &self,
+    publish_items: Vec<PublishCollabItem<serde_json::Value, Vec<u8>>>,
+    workspace_id: &Uuid,
+    user_uuid: &Uuid,
+  ) -> Result<(), AppError> {
+    let publish_items_batch_size = publish_items.len() as i64;
+    for publish_item in &publish_items {
+      check_collab_publish_name(publish_item.meta.publish_name.as_str())?;
+      let object_key = get_collab_s3_key(workspace_id, &publish_item.meta.view_id);
+      let data = publish_item.data.clone();
+      let bucket_client = self.bucket_client.clone();
+      let metrics = self.metrics.clone();
+      tokio::spawn(async move {
+        let result = bucket_client.put_blob(&object_key, &data).await;
+        if let Err(err) = result {
+          debug!("Failed to publish collab to S3: {}", err);
+        } else {
+          metrics.incr_success_write_count(1);
+        }
+      });
+    }
+    let result =
+      insert_or_replace_publish_collabs(&self.pg_pool, workspace_id, user_uuid, publish_items)
+        .await;
+    if result.is_err() {
+      self
+        .metrics
+        .incr_failure_write_count(publish_items_batch_size);
+    } else {
+      self
+        .metrics
+        .incr_fallback_write_count(publish_items_batch_size);
+    }
+    result
+  }
+
+  async fn get_collab_metadata(
+    &self,
+    publish_namespace: &str,
+    publish_name: &str,
+  ) -> Result<serde_json::Value, AppError> {
+    let metadata =
+      select_publish_collab_meta(&self.pg_pool, publish_namespace, publish_name).await?;
+    Ok(metadata)
+  }
+
+  async fn get_collab_with_view_metadata_by_view_id(
+    &self,
+    view_id: &Uuid,
+  ) -> Result<Option<(PublishViewMetaData, Vec<u8>)>, AppError> {
+    let result = select_published_metadata_for_view_id(&self.pg_pool, view_id).await?;
+    match result {
+      Some((workspace_id, js_val)) => {
+        let metadata = serde_json::from_value(js_val)?;
+        let object_key = get_collab_s3_key(&workspace_id, view_id);
+        match self.bucket_client.get_blob(&object_key).await {
+          Ok(resp) => {
+            self.metrics.incr_success_read_count(1);
+            Ok(Some((metadata, resp.to_blob())))
+          },
+          Err(_) => {
+            let result = match select_published_data_for_view_id(&self.pg_pool, view_id).await? {
```

Comment:
When sending collab state to S3, are we removing it from Postgres? Otherwise, I think we should check if we have Error::NotFound here. Without it, over transition period from Postgres&rarr;S3, we might get stale state from postgres.

Reply:
For now, the collab state are written to both postgres and s3 on every publish, so both s3 and postgres will have the same data.